src/clustering_quality.jl

```diff
@@ -0,0 +1,316 @@
+    global_center = vec(mean(data, dims=2))
+    center_distances = colwise(metric, centers, global_center)
+    outer_inertia = length.(cluster_samples) ⋅ center_distances
+
+    inner_inertia = _inner_inertia(metric, data, centers, assignments)
+
+    return (outer_inertia / inner_inertia) * (n - k) / (k - 1)
+end
+
+function  _cluquality_calinski_harabasz(
+        metric::SemiMetric,
+        data::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        assignments::Union{AbstractVector{<:Integer}, AbstractMatrix{<:Real}},
+        fuzziness::Union{Real, Nothing}
+)
+    n, k = size(data, 2), size(centers, 2)
+
+    global_center = vec(mean(data, dims=2))
+    center_distances = colwise(metric, centers, global_center)
+
+    outer_inertia = 
+        sum(sum(w^fuzziness for w in w_col) * d
+            for (w_col, d) in zip(eachcol(weights), center_distances)
+        )
+    inner_inertia = _inner_inertia(metric, data, centers, weights, fuzziness)
+
+    return (outer_inertia / inner_inertia) * (n - k) / (k - 1)
+end
+
+function _cluquality_davies_bouldin(
+        metric::SemiMetric,
+        data::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        assignments::AbstractVector{<:Integer},
+)
+    clu_idx = axes(centers, 2)
+    clu_samples = _gather_samples(assignments, length(clu_idx))
+    clu_diams = [mean(colwise(metric, view(data, :, samples), view(centers, :, clu)))
+                 for (clu, samples) in zip(clu_idx, clu_samples)]
+    center_dists = pairwise(metric, centers, dims=2)
+
+    DB = mean(
+        maximum(@inbounds (cluster_diameters[j₁] + cluster_diameters[j₂]) / center_distances[j₁,j₂] for j₂ in c_idx if j₂ ≠ j₁)
+            for j₁ in c_idx
+    )
+    return  DB
+end
+
+
+# Xie-Beni index
+
+function _cluquality_xie_beni(
+        metric::SemiMetric,
+        data::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        assignments::Union{AbstractVector{<:Integer}, AbstractMatrix{<:Real}},
+        fuzziness::Union{Real, Nothing}
+)
+    n, k = size(data, 2), size(centers,2)
+
+    inner_intertia  = _inner_inertia(metric, data, centers, assignments)
+
+    center_distances = pairwise(metric,centers)
+    min_center_distance = minimum(center_distances[j₁,j₂] for j₁ in 1:k for j₂ in j₁+1:k)
+    
+    return inner_intertia / (n * min_center_distance)
+end
+
+function _cluquality_xie_beni(
+        data::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        weights::AbstractMatrix{<:Real},
+        fuzziness::Real,
+        metric::SemiMetric=SqEuclidean()
+    )
+
+    n, k = size(data, 2), size(centers, 2)
+
+    inner_intertia = _inner_inertia(metric, data, centers, weights, fuzziness)
+
+    center_distances = pairwise(metric, centers, dims=2)
+    min_center_distance = minimum(center_distances[j₁,j₂] for j₁ in 1:k for j₂ in j₁+1:k)
+    return inner_intertia / (n * min_center_distance)
+end
```

Comment:
We don't need this method: the code is exactly the same as the one above:
 * for hard clustering `assignments` is the vector of integer cluster indices for each sample, and `fuzziness = nothing`
 * for fuzzy clustering `assignments` is the matrix of sample-to-cluster weights, and `fuzziness` is a real number
 
 So depending on the type of `assignments` and `fuzziness` the correct version of the `inner_inertia` is called.

Reply:
It was unintended, I made a bad merge.